test/OAuth2ManagerTests/OAuth2APITests.cpp

```diff
@@ -0,0 +1,1225 @@
+                {
+                    if (codeChallenge != codeVerifier)
+                    {
+                        assignInvalidRequestError(L"Code verifier does not match the expected value");
+                    }
+                }
+            }
+            else if (grantType == L"client_credentials")
+            {
+                if (expectedGrantType != L"client")
+                {
+                    assignMismatchedArgsError(L"grant_type", expectedGrantType, grantType);
+                }
+
+                checkUnexpectedArg(L"code", code);
+                checkUnexpectedArg(L"code_verifier", codeVerifier);
+                checkUnexpectedArg(L"refresh_token", refreshToken);
+            }
+            else if (grantType == L"refresh_token")
+            {
+                if (expectedGrantType != L"refresh")
+                {
+                    assignMismatchedArgsError(L"grant_type", expectedGrantType, grantType);
+                }
+                else if (refreshToken != refresh_token_old)
+                {
+                    assignMismatchedArgsError(L"refresh_token", refresh_token_old, refreshToken);
+                }
+
+                checkUnexpectedArg(L"code", code);
+                checkUnexpectedArg(L"code_verifier", codeVerifier);
+            }
+            else if (grantType == extension_grant_uri)
+            {
+                if (expectedGrantType != L"extension")
+                {
+                    assignMismatchedArgsError(L"grant_type", expectedGrantType, grantType);
+                }
+
+                checkUnexpectedArg(L"code", code);
+                checkUnexpectedArg(L"code_verifier", codeVerifier);
+                checkUnexpectedArg(L"refresh_token", refreshToken);
+            }
+            else
+            {
+                assignInvalidRequestError(L"Unrecognized grant type '"s + grantType + L"'");
+            }
+
+            if (errorString.empty())
+            {
+                // NOTE: All responses are the same
+                responseJson = L"{\"access_token\":\"";
+                responseJson += json_escaped_token;
+                responseJson += L"\",\"token_type\":\"Bearer\",\"expires_in\":3600,\"refresh_token\":\"";
+                responseJson += json_escaped_refresh_token;
+                responseJson += L"\"";
+                if (scope.empty())
+                {
+                    responseJson += L",\"scope\":\"all\"";
+                }
+                responseJson += L"}";
+            }
+            else
+            {
+                responseJson = L"{\"error\":\"" + errorString + L"\",\"error_description\":\"" + errorMessage +
+                    L"\",\"error_uri\":\"" + error_uri + L"\",\"" + additional_param_key + L"\":\"" +
+                    additional_param_value + L"\"}";
+            }
+
+            WINRT_ASSERT(!responseJson.empty());
+
+            HTTP_RESPONSE response = {};
+            response.StatusCode = 200;
```

Comment:
Whether `StatusCode` and 'pReason` should be different in case if there is error?

Reply:
No, in the UT we are testing the scenario where server status code is success but response json contains error message and we are validating the same.